library/core/src/ptr/mod.rs

```diff
@@ -537,7 +537,10 @@ pub unsafe fn drop_in_place<T: ?Sized>(to_drop: *mut T) {
 #[rustc_allow_const_fn_unstable(ptr_metadata)]
 #[rustc_diagnostic_item = "ptr_null"]
 pub const fn null<T: ?Sized + Thin>() -> *const T {
-    from_raw_parts(invalid(0), ())
+    // Use transmute instead of casting so that Miri knows that the pointer is invalid.
+    // SAFETY: on all current platforms, usize and pointers have the same layout,
+    // and the validity invariant of pointers is the same as that of integers
+    unsafe { mem::transmute(0usize) }
```

Comment:
If this is just a temporary patch to work around ecosystem issues, as @saethlin put it, then I think I'd prefer to keep using `invalid(0)` here, and constrain the hack (working around codegen not reflecting our intended transmute semantics) to `fn invalid`.

Reply:
I was assuming `invalid` would be `transmute(0usize).with_addr(...)`, with comments explaining what happens.

That keeps all hacks confined to one function.

Reply:
Changed it!
@rustbot ready